File: backend/app/schemas/activity.py
import strawberry
from datetime import datetime
from typing import Optional, List
from dataclasses import field
from bson.objectid import ObjectId
from app.schemas.organization import Organization
from app.config.database import organizations

@strawberry.type
class UserDetails:
    firstName: str
    lastName: str
    email: str
    role: str

@strawberry.type
class Activity:
    id: str
    action: str
    userDetails: Optional[UserDetails]
    organization: Optional[Organization]
    createdAt: datetime
    updatedAt: datetime

    @classmethod
    async def from_db(cls, activity_data: dict, organization: Optional[Organization] = None) -> "Activity":
        """Convert DB activity to Activity type"""
        # Create UserDetails from embedded data
        user_details = None
        if "userDetails" in activity_data:
            user_details = UserDetails(
                firstName=activity_data["userDetails"]["firstName"],
                lastName=activity_data["userDetails"]["lastName"],
                email=activity_data["userDetails"]["email"],
                role=activity_data["userDetails"]["role"]
            )

        # Use provided organization if available, otherwise fetch
        org_obj = organization
        if not org_obj and "organizationId" in activity_data:
            org_data = await organizations.find_one({"_id": activity_data["organizationId"]})
            if org_data:
                org_obj = await Organization.from_db(org_data)

        return cls(
            id=str(activity_data["_id"]),
            action=activity_data["action"],
            userDetails=user_details,
            organization=org_obj,
            createdAt=activity_data["createdAt"],
            updatedAt=activity_data["updatedAt"]
        )

@strawberry.type
class ActivityResponse:
    success: bool
    message: str
    activity: Optional[Activity] = None

@strawberry.type
class ActivitiesResponse:
    success: bool
    message: str
    activities: List[Activity] = field(default_factory=list)
    totalCount: int = 0

File: frontend/graphql/activity.ts
import { gql } from "@apollo/client";

export const GET_ACTIVITIES = gql`
  query GetActivities($organizationId: String!, $limit: Int, $skip: Int) {
    activities(organizationId: $organizationId, limit: $limit, skip: $skip) {
      success
      message
      activities {
        id
        action
        userDetails {
          firstName
          lastName
          email
          role
        }
        organization {
          id
          name
        }
        createdAt
        updatedAt
      }
      totalCount
    }
  }
`;
export const GET_ACTIVITY = gql`
  query GetActivity($id: String!) {
    activity(id: $id) {
      success
      message
      activity {
        id
        action
        userDetails {
          firstName
          lastName
          email
          role
        }
        organization {
          id
          name
        }
        createdAt
        updatedAt
      }
    }
  }
`;
export const CREATE_ACTIVITY = gql`
  mutation CreateActivity($organizationId: String!, $action: String!) {
    createActivity(organizationId: $organizationId, action: $action) {
      success
      message
      activity {
        id
        action
        userDetails {
          firstName
          lastName
          email
          role
        }
        organization {
          id
          name
        }
        createdAt
        updatedAt
      }
    }
  }
`;
export const DELETE_ACTIVITY = gql`
  mutation DeleteActivity($id: String!) {
    deleteActivity(id: $id) {
      success
      message
      activity {
        id
        action
        userDetails {
          firstName
          lastName
          email
          role
        }
        organization {
          id
          name
        }
        createdAt
        updatedAt
      }
    }
  }
`;
export const CLEAR_OLD_ACTIVITIES = gql`
  mutation ClearOldActivities($days: Int) {
    clearOldActivities(days: $days) {
      success
      message
      activity {
        id
      }
    }
  }
`;
